tests/e2e/ctl_v3_auth_cluster_test.go

```diff
@@ -0,0 +1,130 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"sync"
+	"testing"
+)
+
+func TestAuthCluster(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := &e2e.EtcdProcessClusterConfig{
+		ClusterSize:   1,
+		InitialToken:  "new",
+		SnapshotCount: 2,
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	epc, err := e2e.NewEtcdProcessCluster(ctx, t, cfg)
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("could not close test cluster (%v)", err)
+		}
+	}()
+
+	epcClient := epc.Client()
+	createUsers(ctx, t, epcClient)
+
+	if err := epcClient.AuthEnable(ctx); err != nil {
+		t.Fatalf("could not enable Auth: (%v)", err)
+	}
+
+	numKeys := 10
+	testUserClientOpts := e2e.WithAuth("test", "testPassword")
+	rootUserClientOpts := e2e.WithAuth("root", "rootPassword")
+	writeKeys(ctx, epc.Client(testUserClientOpts), numKeys, "test")
+
+	if err := epc.StartNewProc(ctx, t, rootUserClientOpts); err != nil {
+		t.Fatalf("could not start second etcd process (%v)", err)
+	}
+
+	writeKeys(ctx, epc.Client(testUserClientOpts), numKeys, "test_two_nodes")
```

Comment:
Shouldn't you write keys with a new client which connects to the new member?

Reply:
Well, the issue looks different if write hits leader or follower.
In one case write is successful, but not replicated correctly (state is corrupted), in another - client receives permission error. 
To verify both cases I can write to both endpoints and verify no errors and HashKV returns correct state. What do you think?

Reply:
It has nothing to do with "leader" vs "follower". The client connecting to the new member will always fail to read or write data without the fix in this PR, because it doesn't refresh the user data when recovering from a snapshot, accordingly will fail to apply data. 

Suggestion:
1. Comment out the code you fix in this PR. It is line 389 in [server/auth/store.go](https://github.com/etcd-io/etcd/pull/14574/files#diff-42ee8558bd5725beaff504d96ada7c465e0df6bb7cc57c445367e5f1e8783b88);
2. Update the e2e test per my suggestion; then the case should be always failing.
3. Add the code (commented out at first step) back, then the test case should can succeed. 

Of course, you can also manually verify this.

Reply:
Testing both endpoints look good to me.

Reply:
updated test to remove hashkv check.

Reply:
nvm, I see your suggestion below.